### PR TITLE
Fix file reloading in dmypy with --export-types

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/15794

Unfortunately, this requires to pass `--export-types` to `dmypy run`  if one wants to inspect a file that was previously kicked out of the build.

### CODE REVIEW

File: mypy/dmypy_server.py [hunk 1]
```diff
@@ -557,6 +572,14 @@ def fine_grained_increment(
             # Use the remove/update lists to update fswatcher.
             # This avoids calling stat() for unchanged files.
             changed, removed = self.update_changed(sources, remove or [], update or [])
+        if explicit_export_types:
+            # If --export-types is given, we need to force full re-checking of all
+            # explicitly passed files, since we need to visit each expression.
+            changed += [
+                (bs.module, bs.path)
+                for bs in sources
+                if bs.path and (bs.module, bs.path) not in changed
```
Comment (JukkaL) [inline mypy/dmypy_server.py:581]: This is quadratic, and it could be slow if there are many files and many that are changed. Maybe construct a temporary set from `changed` and use it in the `not in changed` check?

File: mypy/dmypy_server.py [hunk 2]
```diff
@@ -601,6 +626,13 @@ def fine_grained_increment_follow_imports(self, sources: list[BuildSource]) -> l
         changed, new_files = self.find_reachable_changed_modules(
             sources, graph, seen, changed_paths
         )
+        if explicit_export_types:
+            # Same as in fine_grained_increment().
+            changed += [
+                (bs.module, bs.path)
+                for bs in sources
+                if bs.path and (bs.module, bs.path) not in changed
```
Comment (JukkaL) [inline mypy/dmypy_server.py:634]: Similar to above. Maybe move this to a helper function to avoid code duplication?

File: mypy/test/testfinegrained.py [hunk 3]
```diff
@@ -149,6 +149,7 @@ def get_options(self, source: str, testcase: DataDrivenTestCase, build_cache: bo
         options.use_fine_grained_cache = self.use_cache and not build_cache
         options.cache_fine_grained = self.use_cache
         options.local_partial_types = True
+        options.export_types = True
```
Comment (JukkaL) [inline mypy/test/testfinegrained.py:152]: What's the purpose of this change?

File: mypy/test/testfinegrained.py [hunk 4]
```diff
@@ -164,7 +165,7 @@ def get_options(self, source: str, testcase: DataDrivenTestCase, build_cache: bo
         return options
 
     def run_check(self, server: Server, sources: list[BuildSource]) -> list[str]:
-        response = server.check(sources, export_types=True, is_tty=False, terminal_width=-1)
+        response = server.check(sources, export_types=False, is_tty=False, terminal_width=-1)
```
Comment (JukkaL) [inline mypy/test/testfinegrained.py:168]: Hmm why do we have `export_types=False` here, but true above?

File: mypy/dmypy_server.py (hunk 1 above)
Comment (ilevkivskyi) [reply mypy/dmypy_server.py]: Yes, good point, fixed now.

File: mypy/dmypy_server.py (hunk 2 above)
Comment (ilevkivskyi) [reply mypy/dmypy_server.py]: Moved to a helper function.

File: mypy/test/testfinegrained.py (hunk 3 above)
Comment (ilevkivskyi) [reply mypy/test/testfinegrained.py]: Starting the daemon with this option set to `True` is cleaner for `inspect` unit tests, I made this explicit now (so it will not slow down other fine grained tests).

File: mypy/test/testfinegrained.py (hunk 4 above)
Comment (ilevkivskyi) [reply mypy/test/testfinegrained.py:167]: Passing it explicitly on each run will force reloading files. Putting it in options from the start is IMO cleaner.

**Review by JukkaL — APPROVED** [review]
Thanks for the updates!
